jobrunner/k8s/k8s_runner.py

```diff
@@ -0,0 +1,483 @@
+from __future__ import print_function, unicode_literals, division, absolute_import
+
+import hashlib
+import re
+import socket
+import time
+from enum import Enum
+
+from jobrunner import config
+
+from kubernetes import client
+from kubernetes import config as k8s_config
+from typing import Mapping, List, Tuple, Optional
+
+batch_v1 = client.BatchV1Api()
+core_v1 = client.CoreV1Api()
+networking_v1 = client.NetworkingV1Api()
+
+
+def init_k8s_config():
+    global batch_v1, core_v1, networking_v1
+    if config.K8S_USE_LOCAL_CONFIG:
+        # for testing, e.g. run on a local minikube
+        k8s_config.load_kube_config()
+    else:
+        # Use the ServiceAccount in the cluster
+        k8s_config.load_incluster_config()
+    batch_v1 = client.BatchV1Api()
+    core_v1 = client.CoreV1Api()
+    networking_v1 = client.NetworkingV1Api()
+
+
+# TODO to be split into multiple functions: prepare / execute / finalize
+def create_opensafely_job(workspace_name, opensafely_job_id, opensafely_job_name, repo_url, private_repo_access_token, commit_sha, inputs, allow_network_access,
+                          execute_job_image, execute_job_command, execute_job_arg, execute_job_env):
+    """
+    1. create pv and pvc (ws_pvc) for the workspace if not exist
+    2. check if the job exists, skip the job if already created
+    3. create pv and pvc (job_pvc) for the job
+    4. create a k8s job with ws_pvc and job_pvc mounted, this job consists of multiple steps running in multiple containers:
+       1. pre container: git checkout study repo to job volume
+       2. job container: run the opensafely job command (e.g. cohortextractor) on job_volume
+       3. post container: use python re to move matching output files from job volume to ws volume
+    """
+    storage_class = config.K8S_STORAGE_CLASS
+    namespace = config.K8S_NAMESPACE
+    jobrunner_image = config.K8S_JOB_RUNNER_IMAGE
+    size = config.K8S_STORAGE_SIZE
+    whitelist = config.K8S_EXECUTION_HOST_WHITELIST
+    
+    ws_pv = convert_k8s_name(workspace_name, "pv")
+    ws_pvc = convert_k8s_name(workspace_name, "pvc")
+    job_pv = convert_k8s_name(opensafely_job_id, "pv")
+    job_pvc = convert_k8s_name(opensafely_job_id, "pvc")
+    
+    create_pv(ws_pv, storage_class, size)
+    create_pv(job_pv, storage_class, size)
```

Comment:
I'm a bit surprised that this uses the same storage class and size for both PVs.

As I understand it, the workspace PV is your implementation of "long-term storage" and it needs to be something that supports `ReadWriteMany` -- presumably Azure Files. But the job PV is only needed for this job, so it can be much less fancy.

And the workspace PV has to fit the outputs of all actions in this workspace, plus their logs; but the job PV only needs to fit outputs from one job.

Reply:
Yes, you are right. This will be changed.